### PR TITLE
Make removing of connections, sessions, plugins promise-based

### PR DESCRIPTION
**Pull request by vogdb**
As discussed keep `onRemove` and `_remove*` methods. Promisify them.

### CODE REVIEW

**Comment by vogdb** [comment]
@tomaszdurka please review

File: lib/janus/connection.js [hunk 1]
```diff
@@ -107,15 +107,18 @@ JanusConnection.prototype.onTimeout = function(message) {
 };
 
 JanusConnection.prototype.onRemove = function() {
-  this._removeSession();
-  serviceLocator.get('logger').info('Removed ' + this);
+  return this._removeSession().finally(function() {
```
Comment (tomaszdurka) [inline lib/janus/connection.js]: What is the reason to use `finally` instead of `then`? We might miss some exceptions this way.

Comment (vogdb) [inline lib/janus/connection.js]: Well. If we did anything on exceptions, I would have used `then` but we do nothing so, I used `finally`. What shall we do on exceptions? We don't handle them anywhere.

File: test/spec/janus/session.js [hunk 2]
```diff
@@ -103,16 +103,19 @@ describe('Session', function() {
   context('when removes plugin', function() {
     beforeEach(function() {
       session.plugins['plugin-id'] = sinon.createStubInstance(PluginAbstract);
+      session.plugins['plugin-id'].onRemove.returns(Promise.resolve());
```
Comment (tomaszdurka) [inline test/spec/janus/session.js]: I found instant returning resolved/rejected Promise risky :) especially with `returns(Promise.reject())` I got unhandled rejections...

Comment (vogdb) [inline test/spec/janus/session.js]: Here it should be safe

File: lib/janus/connection.js (hunk 1 above)
Comment (tomaszdurka) [inline lib/janus/connection.js]: Well, you cannot know from this method what will be done with the exception. Thing is now we are silently ignoring them. I prefer to have unhandled exception then to at least know about it. Agree?

**Comment by vogdb** [comment]
@tomaszdurka please review

File: lib/janus/connection.js [hunk 3]
```diff
@@ -107,15 +107,18 @@ JanusConnection.prototype.onTimeout = function(message) {
...
+  return this._removeSession().then(function() {
+    serviceLocator.get('logger').info('Removed ' + this);
+  }.bind(this));
 };
 
 JanusConnection.prototype._removeSession = function() {
   if (null !== this.session) {
-    this.session.onRemove();
-    this.session = null;
+    return this.session.onRemove().finally(function() {
```
Comment (tomaszdurka) [inline lib/janus/connection.js]: Still finally. We don't want to catch/ignore exception here as well.

File: lib/janus/plugin/streaming.js [hunk 4]
```diff
@@ -83,14 +83,17 @@ PluginStreaming.prototype.getStream = function() {
 
 PluginStreaming.prototype.removeStream = function() {
   if (this.stream) {
-    serviceLocator.get('streams').remove(this.getStream());
-    this.stream = null;
+    return serviceLocator.get('streams').remove(this.getStream()).finally(function() {
```
Comment (tomaszdurka) [inline lib/janus/plugin/streaming.js]: Same.

File: lib/janus/proxy.js [hunk 5]
```diff
@@ -162,22 +162,21 @@ JanusProxy.prototype.addConnection = function(connection) {
  */
 JanusProxy.prototype.removeConnection = function(connection) {
   if (this.connections[connection.id]) {
-    connection.onRemove();
-    delete this.connections[connection.id];
+    return connection.onRemove().finally(function() {
```
Comment (tomaszdurka) [inline lib/janus/proxy.js]: Same.

File: lib/janus/session.js [hunk 6]
```diff
@@ -86,16 +86,20 @@ Session.prototype.onDetached = function(message) {
  */
 Session.prototype._removePlugin = function(id) {
   if (this.plugins[id]) {
-    this.plugins[id].onRemove();
-    delete this.plugins[id];
+    return this.plugins[id].onRemove().finally(function() {
```
Comment (tomaszdurka) [inline lib/janus/session.js]: Same.

**Comment by vogdb** [comment]
@tomaszdurka please review

File: lib/janus/session.js [hunk 7]
```diff
@@ -86,16 +86,20 @@ Session.prototype.onDetached = function(message) {
...
 };
 
 Session.prototype.onRemove = function() {
-  _.each(this.plugins, function(plugin, id) {
-    this._removePlugin(id);
-  }.bind(this));
-  serviceLocator.get('logger').info('Removed ' + this);
+  var self = this;
+  return Promise.all(_.map(this.plugins, function(plugin, id) {
+    return self._removePlugin(id).reflect();
```
Comment (tomaszdurka) [inline lib/janus/session.js]: Not sure if I understand how reflect works. but I trust it.

**Comment by tomaszdurka** [comment]
Looks good.

**Pull request merged**